PR: https://github.com/odoo/odoo/pull/276163

From: ed906a325fc22cb440eca59c81b844967c7bed17
From: Arnold Moyaux
Date: 2026-08-21 12:29:58

Breaking data model changes scores: del:1 + add:0, change matches:
-    _inherit = 'stock.move'

Total Changes: 29

[REF] stock_account: replay COGS on move's value update

It's the second commit of the stock.move revaluation. In the previous
commit, only the stock.move value was updated. In this commit, we need
to adapt the linked account.move.line. There is no direct link between
a stock.move and the COGS account.move.line. Often it's done through
the sale.order.line or purchase.order.line. Another problem: a COGS
line could be linked to multiple stock.move (for example backorders
before the invoice). And conversely, multiple COGS lines for a single
stock.move (1 delivery, multiple invoices).

The current code checks the difference between the delivered value and
the posted COGS. So it's very volatile and could end up with a lot of
different COGS for the same product.

Conclusion: the current system was very complex and makes it very
difficult to correct COGS entries.

The new system is based on a price unit on an intermediate object (e.g.
sale order). To understand, here is an example in FIFO with both
systems:

SO for 2 units:
a) receipt and bill for 1 unit at $10
b) delivery and invoice for 1 unit
c) receipt and bill for 1 unit at $20
d) delivery and invoice for 1 unit

In the old system, the COGS associated with the invoice in b) was $10
and in d) $20. Now a price unit is evaluated per sale.order.line. So at
b) it's still $10 as before. But at d) it reevaluates the price unit to
$15 and updates both COGS lines to $15.

This way it's easy to recompute a new global price unit and update the
COGS.

Part-of: odoo/odoo#276163
Related: odoo/enterprise#124325
Related: odoo/upgrade#10807
Signed-off-by: Arnold Moyaux (arm) <arm@odoo.com>

================================= pseudo patch: =================================

--- a/addons/pos_mrp/models/__init__.py
+++ b/addons/pos_mrp/models/__init__.py
@@ -3,5 +3,4 @@
 
 from . import pos_order
 from . import pos_order_line
-from . import stock_move
 from . import stock_picking

--- a/addons/pos_mrp/models/pos_order_line.py
+++ b/addons/pos_mrp/models/pos_order_line.py
@@ -15,12 +15,4 @@ class PosOrderLine(models.Model):
         # Get a flat list of all bom_line_ids
         bom_line_ids = [item.id for x in boms for item in x[0].bom_line_ids if set(item.bom_product_template_attribute_value_ids.ids).issubset(product.product_template_variant_value_ids.ids)]
         ml_product_to_consider = (product.bom_ids and [comp[0].product_id.id for comp in components]) or [product.id]
-        return stock_moves.filtered(lambda ml: ml.product_id.id in ml_product_to_consider and (ml.bom_line_id.id in bom_line_ids)).with_context(pos_order_line_id=self.id, bom_id=bom.id)
-
-    def _get_product_cost_with_moves(self, moves):
-        self.ensure_one()
-        product = self.product_id
-        bom = self.env['mrp.bom']._bom_find(product, company_id=self.company_id.id, bom_type='phantom').get(product)
-        if bom:
-            return moves._get_kit_price_unit(product, bom, self.qty)
-        return super()._get_product_cost_with_moves(moves)
+        return stock_moves.filtered(lambda ml: ml.product_id.id in ml_product_to_consider and (ml.bom_line_id.id in bom_line_ids))

--- a/addons/pos_mrp/models/stock_move.py
+++ b/None
@@ -1,18 +0,0 @@
-# Part of Odoo. See LICENSE file for full copyright and licensing details.
-
-from odoo import models
-
-
-class StockMove(models.Model):
-    _inherit = 'stock.move'
-
-    def _get_price_unit(self):
-        pos_order_line_id = self.env.context.get('pos_order_line_id')
-        bom_id = self.env.context.get('bom_id')
-        if not pos_order_line_id or not bom_id:
-            return super()._get_price_unit()
-        pos_order_line = self.env['pos.order.line'].browse(pos_order_line_id)
-        bom = self.env['mrp.bom'].browse(bom_id)
-        if pos_order_line and bom:
-            return self._get_kit_price_unit(pos_order_line.product_id, bom, pos_order_line.qty)
-        return super()._get_price_unit()
